README.md
[![Build Status](https://travis-ci.org/pokutuna/p5-App-CpanfileSlipstop.svg?branch=master)](https://travis-ci.org/pokutuna/p5-App-CpanfileSlipstop) [![Coverage Status](https://img.shields.io/coveralls/pokutuna/p5-App-CpanfileSlipstop/master.svg?style=flat)](https://coveralls.io/r/pokutuna/p5-App-CpanfileSlipstop?branch=master) [![MetaCPAN Release](https://badge.fury.io/pl/App-CpanfileSlipstop.svg)](https://metacpan.org/release/App-CpanfileSlipstop)
# NAME

cpanfile-slipstop - write installed module versions back to cpanfile

# SYNOPSIS

    # update moduels & write versions from cpanfile.snapshot to cpanfile
    > carton update
    > cpanfile-slipstop

    # write module versions as 'minimum'
    > cpanfile-slipstop --stopper=minimum

    # only see versions to write
    > cpanfile-slipstop --dry-run

    # remove current version specification from cpanfile
    > cpanfile-slipstop --remove

# OPTIONS

    --stopper=identifier (default: exact)
        type of version constraint
            exact   : '== 1.00'
            minimum : '1.00' (same as >= 1.00)
            maximum : '<= 1.00'
    --dry-run
        do not save to cpanfile
    --with-core
        write core module versions
    --silent
        stop to output versions
    --cpanfile=path (default: ./cpanfile)
    --snapshot=path (default: ./cpanfile.snapshot)
        specify cpanfile and cpanfile.snapshot location
    --remove
        delete all version specifications from cpanfile
    -h, --help
        show this help

# DESCRIPTION

`cpanfile-slipstop` is a support tool for more definite and safety version bundling on [cpanfile](https://metacpan.org/pod/cpanfile) and [Carton](https://metacpan.org/pod/Carton).

The `carton install` command only check to satisfy version specifications in cpanfile. The saved versions in cpanfile.snapshot is not referred until you need to install modules. This sometimes causes confusion and version discrepancy by environments. This tool write versions snapshot to cpanfile to fix module versions.

# SEE ALSO

[Carton](https://metacpan.org/pod/Carton), [Module::CPANfile](https://metacpan.org/pod/Module::CPANfile), [CPAN::Meta::Requirements](https://metacpan.org/pod/CPAN::Meta::Requirements)

# LICENSE

Copyright (C) pokutuna.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

pokutuna <popopopopokutuna@gmail.com>


Changes
Revision history for Perl extension App-CpanfileSlipstop

{{$NEXT}}

0.02 2018-10-12T07:58:44Z
    - fix handling double quotes

0.01 2018-10-12T03:43:21Z

    - original version


lib/App/CpanfileSlipstop/CLI.pm
package App::CpanfileSlipstop::CLI;
use strict;
use warnings;

use Carton::Snapshot;
use Getopt::Long qw/:config posix_default no_ignore_case gnu_compat bundling/;
use Module::CPANfile;
use Pod::Find qw(pod_where);
use Pod::Usage qw(pod2usage);

use App::CpanfileSlipstop::Resolver;
use App::CpanfileSlipstop::Writer;

sub new {
    my ($class) = @_;

    # defaults
    bless +{
        cpanfile  => './cpanfile',
        snapshot  => './cpanfile.snapshot',
        stopper   => 'exact', # or minimum, maximmu
        dry_run   => 0,
        with_core => 0,
        silent    => 0,

        remove    => 0,
        help      => 0,
    }, $class;
}

sub run {
    my ($self, @argv) = @_;

    $self->parse_options(@argv);

    return $self->cmd_help   if $self->{help};
    return $self->cmd_remove if $self->{remove};

    return $self->cmd_feedback;
}

sub parse_options {
    my ($self, @argv) = @_;

    GetOptions(
        'cpanfile=s' => \($self->{cpanfile}),
        'snapshot=s' => \($self->{snapshot}),
        'stopper=s'  => \($self->{stopper}),
        'dry-run'    => \($self->{dry_run}),
        'with-core'  => \($self->{with_core}),
        'silent'     => \($self->{silent}),
        'remove'     => \($self->{remove}),
        'help|h'     => \($self->{help}),
    );
}

sub cmd_feedback {
    my ($self) = @_;

    my $cpanfile = Module::CPANfile->load($self->{cpanfile});
    my $snapshot = Carton::Snapshot->new(path => $self->{snapshot});
    $snapshot->load;

    my $resolver =  App::CpanfileSlipstop::Resolver->new(
        cpanfile  => $cpanfile,
        snapshot  => $snapshot,
        with_core => $self->{with_core},
    );
    $resolver->read_cpanfile_requirements;
    $resolver->merge_snapshot_versions($self->versioning_method, $self->{with_core});

    my $writer = App::CpanfileSlipstop::Writer->new(
        cpanfile_path => $self->{cpanfile},
        dry_run       => $self->{dry_run},
    );
    $writer->set_versions(
        sub { $resolver->get_version_range($_[0]) },
        sub { !$self->{silent} && $self->log(@_) },
    );

    return 0;
}

sub cmd_remove {
    my ($self) = @_;

    my $cpanfile = Module::CPANfile->load($self->{cpanfile});

    my $writer = App::CpanfileSlipstop::Writer->new(
        cpanfile_path => $self->{cpanfile},
        dry_run       => $self->{dry_run},
    );
    $writer->remove_versions(
        sub { !$self->{silent} && $self->log(@_) },
    );

    return 0;
}

sub cmd_help {
    my ($self) = @_;

    pod2usage(
        -input   => pod_where({ -inc => 1 }, 'App::CpanfileSlipstop'),
        -exitval => 0,
    );
}

sub versioning_method {
    my ($self) = @_;

    return +{
        minimum => 'add_minimum',
        maximum => 'add_maximum',
        exact   => 'exact_version',
    }->{$self->{stopper}};
}

sub log {
    my ($self, $log) = @_;

    return if ($log->{before} || '') eq ($log->{after} || '');

    my $quote = $log->{quote};
    print sprintf "%s: %s -> %s\n",
        $log->{module},
        $log->{before} ? sprintf("$quote%s$quote", $log->{before}) : '(unspecified)',
        $log->{after}  ? sprintf("$quote%s$quote", $log->{after})  : '(unspecified)';
}

1;


lib/App/CpanfileSlipstop/Writer.pm
package App::CpanfileSlipstop::Writer;
use strict;
use warnings;

use List::Util qw(first);
use PPI::Document;
use PPI::Find;

sub new {
    my ($class, %opts) = @_;

    bless +{
        cpanfile_path => $opts{cpanfile_path},
        dry_run       => $opts{dry_run},
    }, $class;
}

sub cpanfile_path { $_[0]->{cpanfile_path}   }
sub dry_run       { $_[0]->{dry_run} ? 1 : 0 }

my $statement_finder = sub {
    my (undef, $elem) = @_;

    return $elem->isa('PPI::Statement') && $elem->schild(0)->content eq 'requires';
};

sub set_versions {
    my ($self, $version_getter, $logger) = @_;

    my $doc = PPI::Document->new($self->cpanfile_path);
    my $requirements = $doc->find($statement_finder);

    for my $statement (@$requirements) {
        my ($type, $module, @args) = $statement->schildren;

        my $version_range = $version_getter->($module->string);
        next unless $version_range;

        my @words = grep {
            !($_->isa('PPI::Token::Operator') || $_->content eq ';');
        } @args;

        if (@words % 2 == 0) {
            # insert VERSION
            # - requires MODULE;
            # - requries MODULE, KEY => VALUE;
            $self->insert_version($module, $version_range);
            $logger->({
                type   => 'insert',
                module => $module->literal,
                before => undef,
                after  => $version_range,
                quote  => quote($module),
            });
        } else {
            # replace VERSION
            # - requries MODULE, VERSION;
            # - requries MODULE, VERSION, KEY => VALUE;
            my $current_version = $words[0];
            $self->replace_version($module, $current_version, $version_range);
            $logger->({
                type   => 'replace',
                module => $module->literal,
                before => $current_version->literal,
                after  => $version_range,
                quote  => quote($module),
            });
        }
    }

    $self->writedown_cpanfile($doc);
}

sub remove_versions {
    my ($self, $logger) = @_;

    my $doc = PPI::Document->new($self->cpanfile_path);
    my $requirements = $doc->find($statement_finder);

    for my $statement (@$requirements) {
        my ($type, $module, @args) = $statement->schildren;

        my @words = grep {
            !($_->isa('PPI::Token::Operator') || $_->content eq ';');
        } @args;

        if (@words %2 == 1) {
            my ($op, $version) = @args;

            # collect whitespaces between MODULE and VERSION
            my $whitespaces = [];
            my $token = $op->next_sibling;
            while ($token && $token->isa('PPI::Token::Whitespace')) {
                push @$whitespaces, $token;
                $token = $token->next_sibling;
            }

            $op->remove;
            $_->remove for @$whitespaces;
            $version->remove;

            $logger->({
                type   => 'delete',
                module => $module->string,
                before => $version->string,
                after  => undef,
                quote  => quote($module),
            });
        }
    }

    $self->writedown_cpanfile($doc);
}

sub writedown_cpanfile {
    my ($self, $ppi_doc) = @_;

    return if $self->dry_run;

    open my $out, ">", $self->cpanfile_path
        or die sprintf('%s, %s', $self->cpanfile_path, $!);
    print $out $ppi_doc->serialize;
    close $out;
}

sub insert_version {
    my ($self, $module_elem, $version_range) = @_;

    my $quote = quote($module_elem);
    $module_elem->__insert_after(PPI::Token->new(qq{, $quote$version_range$quote}));
}

sub replace_version {
    my ($self, $module_elem, $version_elem, $version_range) = @_;

    my $quote = quote($module_elem);

    # The giving version on cpanfile must be a string or number for preventing to replace expressions.
    return if !($version_elem->isa('PPI::Token::Quote') || $version_elem->isa('PPI::Token::Number'));

    my $prev_token = $version_elem->previous_sibling;
    $version_elem->remove;
    $prev_token->__insert_after(PPI::Token->new(qq{$quote$version_range$quote}));
}

sub quote {
    my ($elem) = @_;

    return $elem->isa('PPI::Token::Quote::Single') ? "'" : '"';
}

1;


lib/App/CpanfileSlipstop/Resolver.pm
package App::CpanfileSlipstop::Resolver;
use strict;
use warnings;

use CPAN::Meta::Requirements;

sub new {
    my ($class, %args) = @_;

    my $self = bless +{
        reqs      => CPAN::Meta::Requirements->new,
        cpanfile  => $args{cpanfile}, # Module::CPANfile
        snapshot  => $args{snapshot}, # Carton::Snapshot
    }, $class;

    return $self;
}

sub reqs     { $_[0]->{reqs}     }
sub cpanfile { $_[0]->{cpanfile} }
sub snapshot { $_[0]->{snapshot} }

sub read_cpanfile_requirements {
    my ($self) = @_;

    for my $phase (qw(configure build runtime test develop)) {
        $self->reqs->add_requirements(
            $self->cpanfile->prereqs->requirements_for($phase, 'requires')
        );
    }
    # `carton install` only treats 'requires'.
    # https://metacpan.org/source/MIYAGAWA/Carton-v1.0.34/lib/Carton/CPANfile.pm#L38

    return;
}

sub merge_snapshot_versions {
    my ($self, $merge_method, $with_core) = @_;

    my $find_method = $with_core ? 'find_or_core' : 'find';

    my $cpanfile_modules = [ keys %{$self->reqs->as_string_hash} ];

    for my $module (@$cpanfile_modules) {
        next if $self->ignore_module($module); # skip modules url specified

        my $installed_dist = $self->snapshot->$find_method($module);
        if ($installed_dist) {
            my $version = $installed_dist->version_for($module);
            $self->reqs->$merge_method($module, $version) if $version;
        }
    }
}

sub ignore_module {
    my ($self, $module) = @_;

    # ignore like this.
    #
    #   requires 'Class::Enumemon',
    #       mirror => 'https://cpan.metacpan.org/',
    #       dist   => 'POKUTUNA/Class-Enumemon-0.01.tar.gz';

    my $opts = $self->cpanfile->options_for_module($module);
    return 1 if $opts->{dist};
    return 1 if $opts->{url};
    return 0;
}

sub get_version_range {
    my ($self, $module) = @_;

    my $version_range = $self->reqs->requirements_for_module($module);

    # Remove noisy '>= 0'.
    # This causes when setting version by "add_maximum"
    $version_range =~ s/\A>= 0, //;

    return undef if !$version_range || $version_range eq '0';
    return $version_range;
}

1;


lib/App/CpanfileSlipstop.pm
package App::CpanfileSlipstop;
use 5.008001;
use strict;
use warnings;

our $VERSION = "0.01";

1;

__END__

=encoding utf-8

=head1 NAME

cpanfile-slipstop - write installed module versions back to cpanfile

=head1 SYNOPSIS

  # update moduels & write versions from cpanfile.snapshot to cpanfile
  > carton update
  > cpanfile-slipstop

  # write module versions as 'minimum'
  > cpanfile-slipstop --stopper=minimum

  # only see versions to write
  > cpanfile-slipstop --dry-run

  # remove current version specification from cpanfile
  > cpanfile-slipstop --remove

=head1 OPTIONS

  --stopper=identifier (default: exact)
      type of version constraint
          exact   : '== 1.00'
          minimum : '1.00' (same as >= 1.00)
          maximum : '<= 1.00'
  --dry-run
      do not save to cpanfile
  --with-core
      write core module versions
  --silent
      stop to output versions
  --cpanfile=path (default: ./cpanfile)
  --snapshot=path (default: ./cpanfile.snapshot)
      specify cpanfile and cpanfile.snapshot location
  --remove
      delete all version specifications from cpanfile
  -h, --help
      show this help

=head1 DESCRIPTION

C<cpanfile-slipstop> is a support tool for more definite and safety version bundling on L<cpanfile> and L<Carton>.

The C<carton install> command checks only to satisfy version specifications in cpanfile and C<local/>. Even if some module versions are updated in cpanfile.snapshot, the saved versions are not referred until you need to install it. This sometimes causes confusion and version discrepancy between development environment and production. This tool writes versions snapshot to cpanfile to fix module versions.

=head1 SEE ALSO

L<Carton>, L<Module::CPANfile>, L<CPAN::Meta::Requirements>

=head1 LICENSE

Copyright (C) pokutuna.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

pokutuna E<lt>popopopopokutuna@gmail.comE<gt>

=cut
